text/0000-undo-universal-impl-trait.md

```diff
@@ -0,0 +1,238 @@
+
+And so the feature was stabilized.
+
+This RFC wants to undo this.
+
+## Why is `impl Trait` in argument position wrong?
+
+People out there didn’t really realize what was going on. Complains and confusions on IRC, reddit
+and the Rust forum started to appear, after the stabilization, by different kind of people –
+confusion among newcomers, advanced proficient Rustaceans, etc.. Here are some – non-exhaustive –
+links:
+
+  - https://internals.rust-lang.org/t/how-is-impl-display-different-from-foo-t-display-x-t-t/7566/3
+  - https://www.reddit.com/r/rust/comments/8jfn7z/what_is_the_advantage_of_impl_trait_in_argument/
+  - https://www.reddit.com/r/rust/comments/8io4q6/impl_trait_is_here_now_what/dyu5dlm/
+  - https://github.com/rust-lang/rust/issues/44721#issuecomment-330935175
+
+[RFC 1951] is organized with *arguments* that we’re basing the current document against.
+
+### Argument from learnability
+
+> Argument [here](https://github.com/rust-lang/rfcs/blob/master/text/1951-expand-impl-trait.md#argument-from-learnability)
+
+This argument explains that people who have an understanding of generics would try to use them in
+return position for anonymized types (existentials). This argument then stated it’s confusing
+because the only way to do this with current Rust is to use a complete different mechanism – being
+`impl Trait`. By contrast, if they have first learned `impl Trait` in argument position, trying to
+use it in return position *will just work as intended*.
+
+This argument is twisted along the lines of subjective opinion about learnability. Rust is a
+language with a fabulous set of learning resources (the first and most important one being
+[The Book]). It seems sound to state that **if** people are introduced first to `impl Trait` in
+argument position, it’s likely that they’ll understand the use in return location. However, what if
+they don’t want to express it? Existentials in return locations are already a very specific use case
+of a type system. Having a complete different mechanism is actually a feature, because we are
+talking about something that has no alternative elsewhere in Rust – i.e. pick the type in the
+callee. The initial assumption that having a different mechanism is confusing is wrong because
+`impl Trait` carries its own semantics and because anonymization in return type is not a trivial
+feature people will use without actually wanting it.
+
+Also, people are used to generics. People coming from C++, C#, Java etc. actually know about
+generics. The links listed above prove how confusing it can get to show them a complete, different
+way to use generics (`impl Trait` in argument position is a use of hidden generics, that is,
+anonymized type variables with explicit trait bounds). For people who don’t come from such
+languages or for completely newcomers, we are assuming that generics are too hard to understand at
+first but people will eventually need to learn them, as `impl Trait` in argument position is
```

Comment:
This is no different than lifetime elision. `fn foo(&self)` is strictly less powerful than `fn foo<'a>(&'a self)`.

Reply:
How so?

Reply:
`fn foo(&self, x:  &str) -> &str` vs `fn foo<'a, 'b>(&'a self, x: &'b str) -> &'b str`. You can use the second in more places, e.g. with `'a=lifetime of some object` and `'b = 'static` (or just the lifetime of some longer lived object).